### PR TITLE
convert pumpWebSocketLoop into coros

### PR DESCRIPTION
**Pull request by mikea**
Opens doors for improved error handling according to the comment.

### CODE REVIEW

File: c++/src/kj/compat/http.c++ [hunk 1]
```diff
@@ -3759,34 +3759,33 @@ kj::Own<WebSocket> newWebSocket(kj::Own<kj::AsyncIoStream> stream,
...
       }
-    }
-    KJ_UNREACHABLE;
-  }, [&to](kj::Exception&& e) -> kj::Promise<void> {
-    // We don't know if it was a read or a write that threw. If it was a read that threw, we need
-    // to send a disconnect on the destination. If it was the destination that threw, it
-    // shouldn't hurt to disconnect() it again, but we'll catch and squelch any exceptions.
-    kj::runCatchingExceptions([&to]() { to.disconnect(); });
+      // continue the loop
+    } catch (kj::Exception e) {
```
Comment (danlapid) [inline c++/src/kj/compat/http.c++:3779]: I believe this should be 
```
catch (...) {
    auto e = kj::getCaughtExceptionAsKj();
    ...
```
But if you're only getting the variable to rethrow you can just
```
catch(...) {
    ...
    throw;
}
```

Comment (jclee) [reply c++/src/kj/compat/http.c++:3779]: I guess another possibility would be to put the disconnect code in a `KJ_ON_SCOPE_FAILURE()` block at the top of the scope.

https://github.com/capnproto/capnproto/blob/9ba75a7c/c%2B%2B/src/kj/exception.h#L386-L394

Comment (danlapid) [reply c++/src/kj/compat/http.c++:3779]: Honestly I prefer idiomatic c++ try catch clauses.

File: c++/src/kj/compat/http.c++ [hunk 2]
```diff
@@ -3759,34 +3759,33 @@ kj::Own<WebSocket> newWebSocket(kj::Own<kj::AsyncIoStream> stream,
...
-        return to.send(data)
-            .attach(kj::mv(data))
-            .then([&from,&to]() { return pumpWebSocketLoop(from, to); });
-      }
-      KJ_CASE_ONEOF(close, WebSocket::Close) {
-        // Once a close has passed through, the pump is complete.
-        return to.close(close.code, close.reason)
-            .attach(kj::mv(close));
+  while (true) {
+    try {
```
Comment (kentonv) [inline c++/src/kj/compat/http.c++:3763]: The `try` could be outside the `while` loop. Not sure if it's any more efficient that way... but it's certainly not less efficient.

**Review by danlapid — APPROVED** [review]

**Pull request merged**